src/error.rs

```diff
@@ -0,0 +1,385 @@
+    OutOfMemory,
+
+    #[error("Cost Exceeded")]
+    CostExceeded,
+
+    #[error("Cost must be greater than zero.")]
+    CostBelowZero,
+
+    #[error("Too Many Pairs")]
+    TooManyPairs,
+
+    #[error("Too Many Atoms")]
+    TooManyAtoms,
+
+    #[error("Path Into Atom")]
+    PathIntoAtom,
+
+    #[error("Unknown Softfork Extension")]
+    UnknownSoftforkExtension,
+
+    #[error("Softfork specified cost mismatch")]
+    SoftforkSpecifiedCostMismatch,
+
+    #[error("clvm raise")]
+    Raise(NodePtr),
+
+    #[error("in ((X)...) syntax X must be lone atom")]
+    InPairMustBeLoneAtom(NodePtr),
+
+    #[error("Invalid Nil Terminator")]
+    InvalidNilTerminator(NodePtr),
+
+    #[error("First of non-cons")]
+    FirstOfNonCons(NodePtr),
+
+    #[error("Rest of non-cons")]
+    RestOfNonCons(NodePtr),
+
+    #[error("Division by zero")]
+    DivisionByZero(NodePtr),
+
+    #[error("Divmod by zero")]
+    DivmodByZero(NodePtr),
+
+    #[error("Mod by zero")]
+    ModByZero(NodePtr),
+
+    #[error("ModPow with Negative Exponent")]
+    ModPowNegativeExponent(NodePtr),
+
+    #[error("ModPow with 0 Modulus")]
+    ModPowZeroModulus(NodePtr),
+
+    #[error("Shift too large")]
+    ShiftTooLarge(NodePtr),
+
+    #[error("Value Stack Limit Reached")]
+    ValueStackLimitReached(NodePtr),
+
+    #[error("Environment Stack Limit Reached")]
+    EnvironmentStackLimitReached(NodePtr),
+
+    // Grouped errors
+    #[error("Operator Error: {0}")]
+    Operator(#[from] OperatorError),
+
+    #[error("Allocator Error: {0}")]
+    Allocator(#[from] AllocatorError),
+}
+impl From<std::io::Error> for EvalErr {
+    fn from(_: std::io::Error) -> Self {
+        EvalErr::SerializationError
+    }
+}
+
+impl EvalErr {
+    fn node(&self) -> Option<NodePtr> {
+        match self {
+            EvalErr::Raise(node) => Some(*node),
+            EvalErr::InPairMustBeLoneAtom(node) => Some(*node),
+            EvalErr::InvalidNilTerminator(node) => Some(*node),
+            EvalErr::FirstOfNonCons(node) => Some(*node),
+            EvalErr::RestOfNonCons(node) => Some(*node),
+            EvalErr::DivisionByZero(node) => Some(*node),
+            EvalErr::DivmodByZero(node) => Some(*node),
+            EvalErr::ModByZero(node) => Some(*node),
+            EvalErr::ModPowNegativeExponent(node) => Some(*node),
+            EvalErr::ModPowZeroModulus(node) => Some(*node),
+            EvalErr::ShiftTooLarge(node) => Some(*node),
+            EvalErr::ValueStackLimitReached(node) => Some(*node),
+            EvalErr::EnvironmentStackLimitReached(node) => Some(*node),
+            EvalErr::Operator(op) => OperatorError::node(op),
+            EvalErr::Allocator(alloc) => AllocatorError::node(alloc),
+            _ => None,
+        }
+    }
+    pub fn combined_str(&self) -> String {
+        // This is a convenience function to get the combined string representation of the error
+        match self.node() {
+            Some(node) => format!("{self}: {node:?}"),
```

Comment:
Is the node ever useful for a human? It's an index into an array in the `Allocator`, I don't think it's very useful to debug-print it here. I would expect all uses of `combined_str()` could just use `to_string()` instead.

Reply:
its just to give a more verbose thing for the tests to compare

Reply:
need to check diff and re add nodeptr comparison where i removed it.

Reply:
this is used by wasm as well.